packages/integrations/markdoc/src/runtime.ts

```diff
@@ -0,0 +1,76 @@
+import type { MarkdownHeading } from '@astrojs/markdown-remark';
+import Markdoc, {
+	type RenderableTreeNode,
+	type ConfigType as MarkdocConfig,
+} from '@markdoc/markdoc';
+import type { ContentEntryModule } from 'astro';
+import * as astroDefaultNodes from './nodes/index.js';
+
+export { default as Markdoc } from '@markdoc/markdoc';
+
+export function applyDefaultConfig(
+	config: MarkdocConfig,
+	entry: ContentEntryModule
+): MarkdocConfig {
+	return {
+		...config,
+		variables: {
+			entry,
+			...config.variables,
+		},
+		nodes: {
+			...astroDefaultNodes,
+			...config.nodes,
+		},
+		// TODO: Syntax highlighting
+	};
+}
+
+/**
+ * Get text content as a string from a Markdoc transform AST
+ */
+export function getTextContent(childNodes: RenderableTreeNode[]): string {
+	let text = '';
+	for (const node of childNodes) {
+		if (typeof node === 'string' || typeof node === 'number') {
+			text += node;
+		} else if (typeof node === 'object' && Markdoc.Tag.isTag(node)) {
+			text += getTextContent(node.children);
+		}
+	}
+	return text;
+}
+
+const headingLevels = [1, 2, 3, 4, 5, 6] as const;
+
+/**
+ * Collect headings from Markdoc transform AST
+ * for `headings` result on `render()` return value
+ */
+export function collectHeadings(children: RenderableTreeNode[]): MarkdownHeading[] {
+	let collectedHeadings: MarkdownHeading[] = [];
+	for (const node of children) {
+		if (typeof node !== 'object' || !Markdoc.Tag.isTag(node)) continue;
+
+		if (node.attributes.__collectHeading === true && typeof node.attributes.level === 'number') {
+			collectedHeadings.push({
+				slug: node.attributes.id,
+				depth: node.attributes.level,
+				text: getTextContent(node.children),
+			});
+			continue;
+		}
+
+		for (const level of headingLevels) {
+			if (node.name === 'h' + level) {
+				collectedHeadings.push({
+					slug: node.attributes.id,
+					depth: level,
+					text: getTextContent(node.children),
+				});
+			}
+		}
+		collectedHeadings.concat(collectHeadings(node.children));
```

Comment:
Do we care about the order of how we find the headings? If we don't, I believe we can avoid recursion by using a simple queue

```js
const queue = [...node.children];

while (queue.length > 0) {
	const node = queue.shift();
	// do stuff
	// now process children
	queue.push(...node.children);
}

```

Reply:
Hm, we _do_ want all headings to be in-order by how they appear in the document. I'm a bit confused how we could avoid recursion with a queue here?